### PR TITLE
Fetch item/total counts separate

### LINKED ISSUE #14796: Performance regression in file library on ~5,000,000+ items
### Describe the Bug

TLTR:
from app it takes 60 secs to load /app/files meanwhile API returns /files in 30 ms.
Same for other collections. 3 bottlenecks: sort, filter_count, total_count

I have around 6m files in my directus projects.
Through the app when you nagivate to /files, it makes 2 requests:

This request returns latest files & total_count: 
https://directus.api/files?limit=100&fields[]=id&fields[]=modified_on&fields[]=type&fields[]=title&fields[]=type&fields[]=filesize&sort[]=-uploaded_on&page=1&filter=%7B%22_and%22:[%7B%22type%22:%7B%22_nnull%22:true%7D%7D,%7B%22folder%22:%7B%22_null%22:true%7D%7D]%7D&meta[]=filter_count&meta[]=total_count

Original request takes 30 secs to return on apple m1. When remove all filtersa and ask only /files it takes 30 ms, super fast.
3 things creates bottlenecks:
sort[]=-uploaded_on (adding this increased request time from 30ms to 12secs)
meta[]= filter_count (adding this increased request time from 30ms to 11secs)
meta[]= total_count (adding this increased request time from 30ms to 2.2secs)

Info: filter_count and total_count returns the same value: around 6million.

This request returns only total_count: The request takes 12 secs to return on apple m1.
https://directus.api/files?limit=100&filter=%7B%22_and%22:[%7B%22type%22:%7B%22_nnull%22:true%7D%7D,%7B%22folder%22:%7B%22_null%22:true%7D%7D]%7D&fields[]=id&fields[]=title&fields[]=description&fields[]=tags&fields[]=location&fields[]=storage&fields[]=storage_divider&fields[]=filename_disk&fields[]=filename_download&fields[]=metadata&fields[]=type&fields[]=filesize&fields[]=modified_by&fields[]=modified_on&fields[]=embed&fields[]=uploaded_by&fields[]=uploaded_on&fields[]=folder&fields[]=width&fields[]=height&fields[]=charset&fields[]=duration&fields[]=training&fields[]=inference&fields[]=annotations&fields[]=$thumbnail&sort=-uploaded_on&aggregate=%7B%22count%22:[%22*%22]%7D

This 2nd request seems unnecessary as App already gets total_count from the first request. Also, excluding all fields reduces to time to get from 12 secs to 2 secs for the same requested data. 

As app requests to gets both of these at the same time, it takes around 60 secs to complete, CPU hits 100, everything slows down. This applies 

I do not know what can be done but I know this makes the APP less useful when it takes up-to a minute to load page. 

### To Reproduce

Create millions of items in a collection and try to open a collection from the APP

### Errors Shown

No errors

### What version of Directus are you using?

v9.13.0

### What version of Node.js are you using?

N/A

### What database are you using?

mysql 8

### What browser are you using?

chrome

### How are you deploying Directus?

local

### PR DESCRIPTION
**Pull request by rijkvanzanten**
## Description

Every time the item parameters changed in `use-items`, the `total_count` and `filter_count` were fetched separately. This causes some pretty nasty performance regressions in browsing the app, as you'd have to wait for the counts even if you're doing actions that don't change the counts. This PR resolves that by fetching the total / filtered counts asynchronously, and only updating the counts when needed.

Fixes #14796, fixes ENG-214, fixes ENG-75

## Type of Change

- [x] Bugfix
- [x] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [x] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

**Review by br41nslug — APPROVED** [review]
This does prevent the total count from being updated in the app, in the case where it's updated on the backend, until you've opened another page but that's such a small edge case i doubt most will notice.
On a huge table this makes the App a lot more responsive especially noticeable when searching 🙌

File: packages/shared/src/composables/use-items.ts [hunk 1]
```diff
@@ -205,4 +210,38 @@ export function useItems(collection: Ref<string | null>, query: ComputedQuery, f
 		const endpoint = computed(() => `/utils/sort/${collection.value}`);
 		await api.post(endpoint.value, { item, to });
```
Comment (azrikahar) [inline packages/shared/src/composables/use-items.ts:211]: Unrelated observation here, but maybe we can remove the usage of computed here to make it cleaner?

or if we want to be safer, maybe we can mirror the same return null logic when `collection.value` is not set:

https://github.com/directus/directus/blob/8f5a0723525516f8e9e1c26515f64446b146748f/packages/shared/src/composables/use-items.ts#L41-L46

and prevent this POST request from firing, similar to getItems' approach:

https://github.com/directus/directus/blob/8f5a0723525516f8e9e1c26515f64446b146748f/packages/shared/src/composables/use-items.ts#L103

**Review by azrikahar — COMMENTED** [review]
## Review

FeelsGTM! But I do notice when we're doing batch deletion, the count isn't being updated now. 

Update: same issue applies for batch archiving.

- Before this PR (does still update after batch delete)

    https://user-images.githubusercontent.com/42867097/208884013-8ac6a3f9-5c9b-4788-ba4a-ce2b197fc8d2.mp4

- After this PR (doesn't update anymore after batch delete)

    https://user-images.githubusercontent.com/42867097/208883910-2d866742-7c7a-4510-af45-7e2311aa2ae3.mp4

This is because whenever a layout gets refreshed (such as after batch deletion), `refresh()` does call `getItems()` again, but "deleting item(s)" doesn't prompt it to re-fetch the counts:

https://github.com/directus/directus/blob/cb05cc2812ccc2c39da215183aac49028f39fde5/app/src/layouts/tabular/index.ts#L123-L125

As a quick test, I can confirm exposing and adding `getTotalCount` + `getFilterCount` into this `refresh()` function does seem to resolve this, but there might be a better approach.

## Additional Thoughts

### `itemCount` ref naming

This isn't really related to the changes in this PR, but wondering can/should we rename `itemCount` ref to `filterCount`? 🤔

### Additional aggregate count API calls in Export sidebar

I've also noticed that the getItemCount here in export-sidebar-detail is also firing whenever we navigate to any collection without opening the Export Items drawer, _and_ whenever we add/remove headers in table layout:

https://github.com/directus/directus/blob/cb05cc2812ccc2c39da215183aac49028f39fde5/app/src/views/private/components/export-sidebar-detail.vue#L343-L372

Wondering about the following points:

- Can it be optimized to only be fetched when the drawer is opened? (but then "Exporting all X items" notice may require a loading state for the number of items)

- Would `...exportSettings` in the params be slightly overkill? I believe fields and sort for example is already "cleaned" in the API side when aggregating, so perhaps we can avoid passing it in the request in the first place to make it slightly cleaner.

    https://github.com/directus/directus/blob/8f5a0723525516f8e9e1c26515f64446b146748f/api/src/utils/get-ast-from-query.ts#L53-L59

    https://github.com/directus/directus/blob/8f5a0723525516f8e9e1c26515f64446b146748f/api/src/utils/get-ast-from-query.ts#L94-L98

- seems like it is using `.then` here to get the count compared to approach in this PR, while both does achieve the same result, maybe we should make them consistent, depending on which approach is preferred

but we can also move this `export-sidebar-detail` tangent to a separate discussion/PR if needed.

**Comment by azrikahar** [comment]
Additional note: Took a stab at adding test for this, but had to resort to installing `@vue/test-utils` in shared package to use `flushPromises` so that the computed values are updated, not sure if there's a better way. 

I suspect the reason that they're not updating without `flushPromises`, has something to do with the usage of lodash `throttle()` when fetching items, but using `vi.useFakeTimers()` and advancing the time past the throttle duration doesn't seem to update the computed values either from what I've tried. 

Mocking `throttle()` directly as `vi.fn().mockImplementation(fn => fn)` also isn't really limiting the number of requests either so the test case will be inaccurate, but maybe it's also somewhat related to `fetchOnInit` which seems to be fetching it multiple times and I'm not sure if that's intended, as discussed in https://github.com/directus/directus/discussions/16290#discussioncomment-4055453.

**Comment by br41nslug** [comment]
> But I do notice when we're doing batch deletion, the count isn't being updated now.

Azri brings up a good point when batch deleting (and probably archiving) having the counter off on smaller tables feels like it has a bigger impact than on huge tables. We should make sure that the total counter gets updated when these actions are used. 

Refetching is the quick solution here but for a more performance oriented fix we *could* fix this on the frontend because we know the total count and amount to be removed. (This is however easier said than done in the current structure)

**Comment by rijkvanzanten** [comment]
@azrikahar I agree with all your points, but I feel like most of them are optimizations that are out of scope for this fix in the current setup 😬 I'd rather create a new issue to collect these refactor points, and then do spend a cycle on cleaning it up (and adding the proper tests!) 🙂

File: packages/shared/src/composables/use-items.ts (hunk 1 above)
Comment (rijkvanzanten) [reply packages/shared/src/composables/use-items.ts:224]: Gonna compile this as a larger refactoring exercise 👍🏻

**Pull request merged**